# Uvod v Pandas

## O Pandas-u


**Dimension & Description**



<table class="table table-bordered">
<tbody><tr>
<th style="text-align:center;">Data Structure</th>
<th style="text-align:center;">Dimensions</th>
<th style="text-align:center;">Description</th>
</tr>
<tr>
<td style="text-align:center;">Series</td>
<td style="text-align:center;">1</td>
<td style="text-align:center;">1D labeled homogeneous array, sizeimmutable.</td>
</tr>
<tr>
<td style="text-align:center;">Data Frames</td>
<td style="text-align:center;">2</td>
<td style="text-align:center;">General 2D labeled, size-mutable tabular structure with potentially heterogeneously typed
columns.</td>
</tr>
<tr>
<td style="text-align:center;">Panel</td>
<td style="text-align:center;">3</td>
<td style="text-align:center;">General 3D labeled, size-mutable array.</td>
</tr>
</tbody></table>


## Importing pandas

In [1]:
import pandas as pd

Check the version:

In [2]:
pd.__version__

'0.24.2'

## Reminder about Built-In Documentation



For example, to display all the contents of the pandas namespace, you can type

In [ ]:
#pd.<TAB>
pd.

And to display Pandas's built-in documentation, you can use this:

In [4]:
#pd.DataFrame.max?
pd?

More detailed documentation, along with tutorials and other resources, can be found at http://pandas.pydata.org/.

## Introducing Pandas Objects



In [2]:
import numpy as np
import pandas as pd

### The Pandas Series Object


In [5]:
data = pd.Series([0.25, 0.5, 0.75, 1.01])
data

0    0.25
1    0.50
2    0.75
3    1.01
dtype: float64

In [6]:
type(data)

pandas.core.series.Series

In [7]:
data.values

array([0.25, 0.5 , 0.75, 1.01])

In [8]:
type(data.values) 
#atribute values

numpy.ndarray

In [9]:
data.index

RangeIndex(start=0, stop=4, step=1)

In [10]:
type(data.index)
#atribute index

pandas.core.indexes.range.RangeIndex

In [11]:
data[1]

0.5

In [12]:
data[1:3]

1    0.50
2    0.75
dtype: float64

#### Series as generalized NumPy array


In [16]:
data2 = pd.Series([0.25, 0.5, 0.75, 1.01],
                 index=['a', 'b', 'c', 'd'])
data2

a    0.25
b    0.50
c    0.75
d    1.01
dtype: float64

In [17]:
data2[1]

0.5

In [18]:
data2['b']

0.5

#### Series as specialized dictionary

In [24]:
population_dict = {'California': 383352256,
                  'Texas': 24557785,
                  'New York': 25478962,
                  'Florida': 56555587,
                  'Illinois': 45877896}
population = pd.Series(population_dict)
population


California    383352256
Texas          24557785
New York       25478962
Florida        56555587
Illinois       45877896
dtype: int64

#### Constructing Series objects

`>>> pd.Series(data, index=index)`


[Dokumentacija](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html)

### The Pandas DataFrame Object



In [30]:
area_dict = {'California': 383352256,
             'Texas': 24557785,
             'New York': 25478962,
             'Florida': 56555587,
             'Illinois': 45877896}
area_dict
                              

{'California': 383352256,
 'Texas': 24557785,
 'New York': 25478962,
 'Florida': 56555587,
 'Illinois': 45877896}

In [31]:
area = pd.Series(area_dict)
area

California    383352256
Texas          24557785
New York       25478962
Florida        56555587
Illinois       45877896
dtype: int64

In [33]:
states = pd.DataFrame({'population': population, 'area':area})
states

population       area
California   383352256  383352256
Texas         24557785   24557785
New York      25478962   25478962
Florida       56555587   56555587
Illinois      45877896   45877896

In [34]:
type(states)

pandas.core.frame.DataFrame

In [35]:
states.index

Index(['California', 'Texas', 'New York', 'Florida', 'Illinois'], dtype='object')

In [36]:
states.columns

Index(['population', 'area'], dtype='object')

#### DataFrame as specialized dictionary


In [38]:
pd.DataFrame(population, columns=['population'])

population
California   383352256
Texas         24557785
New York      25478962
Florida       56555587
Illinois      45877896

#### Constructing DataFrame objects



- **From a single Series object**



 - **From a dictionary of Series objects**
 - **From a two-dimensional NumPy array**
 


In [39]:
pd.DataFrame(np.random.rand(3,2),
            columns=['foo', 'bar'],
            index=['a','b','c'])

NameError: name 'np' is not defined

- **From a list of dicts**



In [40]:
import pandas as pd
pd.DataFrame({'a':1, 'b':2}, {'b':3, 'c':4})


a  b
b  1  2
c  1  2

[Dokumentacija](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html)

### The Pandas Index Object



In [ ]:
import pandas as pd

In [42]:
ind = pd.Index([2,3,5,7,11])
ind

Int64Index([2, 3, 5, 7, 11], dtype='int64')

In [ ]:
ind[1]

In [ ]:
ind.size

In [ ]:
ind.shape

In [ ]:
ind.dtype

In [ ]:
ind.dnim

#### Index as immutable array



## Importing Data with Pandas

In [2]:
import pandas as pd
import numpy as np

In [1]:
!head -n 2 data/DATA_fortune_500_2017.csv

In [6]:
import pandas as pd
import numpy as np
f500 = pd.read_csv('data/DATA_fortune_500_2017.csv', index_col=0)

In [7]:
type(f500)

pandas.core.frame.DataFrame

In [8]:
f500.shape

(500, 22)

##Introducing DataFrames



<p><img alt="anatomy of a dataframe" src="https://s3.amazonaws.com/dq-content/291/df_anatomy.svg"></p>



In [9]:
f500.dtypes

Title           object
Website         object
Employees        int64
Sector          object
Industry        object
Hqlocation      object
Hqaddr          object
Hqcity          object
Hqstate         object
Hqzip            int64
Hqtel           object
Ceo             object
Ceo-title       object
Address         object
Ticker          object
Fullname        object
Revenues         int64
Revchange      float64
Profits        float64
Prftchange      object
Assets           int64
Totshequity    float64
dtype: object

In [11]:
f500.head() # default 5 vrstic

Title                           Website  Employees  \
Rank                                                                    
1                Walmart            http://www.walmart.com    2300000   
2     Berkshire Hathaway  http://www.berkshirehathaway.com     367700   
3                  Apple              http://www.apple.com     116000   
4            Exxon Mobil         http://www.exxonmobil.com      72700   
5               McKesson           http://www.mckesson.com      68000   

           Sector                                  Industry  \
Rank                                                          
1       Retailing                     General Merchandisers   
2      Financials  Insurance: Property and Casualty (Stock)   
3      Technology               Computers, Office Equipment   
4          Energy                        Petroleum Refining   
5     Wholesalers                  Wholesalers: Health Care   

             Hqlocation                  Hqaddr         Hqcity Hqstate  Hqzip  \
Rank                                                                            
1       Bentonville, AR     702 S.W. Eighth St.    Bentonville      AR  72716   
2             Omaha, NE         3555 Farnam St.          Omaha      NE  68131   
3         Cupertino, CA         1 Infinite Loop      Cupertino      CA  95014   
4            Irving, TX  5959 Las Colinas Blvd.         Irving      TX  75039   
5     San Francisco, CA              1 Post St.  San Francisco      CA  94104   

      ...                                      Ceo-title  \
Rank  ...                                                  
1     ...  President, Chief Executive Officer & Director   
2     ...             Chairman & Chief Executive Officer   
3     ...             Chief Executive Officer & Director   
4     ...             Chairman & Chief Executive Officer   
5     ...  Chairman, President & Chief Executive Officer   

                                         Address Ticker  \
Rank                                                      
1     702 S.W. Eighth St., Bentonville, AR 72716    WMT   
2               3555 Farnam St., Omaha, NE 68131   BRKA   
3           1 Infinite Loop, Cupertino, CA 95014   AAPL   
4       5959 Las Colinas Blvd., Irving, TX 75039    XOM   
5            1 Post St., San Francisco, CA 94104    MCK   

                     Fullname Revenues Revchange  Profits  Prftchange  Assets  \
Rank                                                                            
1       Wal-Mart Stores, Inc.   485873       0.8  13643.0        -7.2  198825   
2     Berkshire Hathaway Inc.   223604       6.1  24074.0           0  620854   
3                 Apple, Inc.   215639      -7.7  45687.0       -14.4  321686   
4     Exxon Mobil Corporation   205004     -16.7   7840.0       -51.5  330314   
5        McKesson Corporation   192487       6.2   2258.0          53   56563   

     Totshequity  
Rank              
1        77798.0  
2       283001.0  
3       128249.0  
4       167325.0  
5         8924.0  

[5 rows x 22 columns]

In [12]:
f500.head(2) # prve 2 vrstici

Title                           Website  Employees  \
Rank                                                                    
1                Walmart            http://www.walmart.com    2300000   
2     Berkshire Hathaway  http://www.berkshirehathaway.com     367700   

          Sector                                  Industry       Hqlocation  \
Rank                                                                          
1      Retailing                     General Merchandisers  Bentonville, AR   
2     Financials  Insurance: Property and Casualty (Stock)        Omaha, NE   

                   Hqaddr       Hqcity Hqstate  Hqzip  ...  \
Rank                                                   ...   
1     702 S.W. Eighth St.  Bentonville      AR  72716  ...   
2         3555 Farnam St.        Omaha      NE  68131  ...   

                                          Ceo-title  \
Rank                                                  
1     President, Chief Executive Officer & Director   
2                Chairman & Chief Executive Officer   

                                         Address Ticker  \
Rank                                                      
1     702 S.W. Eighth St., Bentonville, AR 72716    WMT   
2               3555 Farnam St., Omaha, NE 68131   BRKA   

                     Fullname Revenues Revchange  Profits  Prftchange  Assets  \
Rank                                                                            
1       Wal-Mart Stores, Inc.   485873       0.8  13643.0        -7.2  198825   
2     Berkshire Hathaway Inc.   223604       6.1  24074.0           0  620854   

     Totshequity  
Rank              
1        77798.0  
2       283001.0  

[2 rows x 22 columns]

In [13]:
f500.tail()  # zadnjih 5 vrstic

Title                      Website  Employees  \
Rank                                                           
496    Michaels Cos.      http://www.michaels.com      31000   
497    Toll Brothers  http://www.tollbrothers.com       4200   
498            Yahoo         http://www.yahoo.com       8500   
499    Vistra Energy  http://www.vistraenergy.com       4431   
500   ABM Industries           http://www.abm.com     110000   

                          Sector                          Industry  \
Rank                                                                 
496                    Retailing        Specialty Retailers: Other   
497   Engineering & Construction                      Homebuilders   
498                   Technology   Internet Services and Retailing   
499                       Energy                            Energy   
500            Business Services  Diversified Outsourcing Services   

         Hqlocation                Hqaddr     Hqcity Hqstate  Hqzip  ...  \
Rank                                                                 ...   
496      Irving, TX  8000 Bent Branch Dr.     Irving      TX  75063  ...   
497     Horsham, PA     250 Gibraltar Rd.    Horsham      PA  19044  ...   
498   Sunnyvale, CA        701 First Ave.  Sunnyvale      CA  94089  ...   
499      Dallas, TX        1601 Bryan St.     Dallas      TX  75201  ...   
500    New York, NY       1 Liberty Plaza   New York      NY  10006  ...   

                                          Ceo-title  \
Rank                                                  
496              Chairman & Chief Executive Officer   
497              Chief Executive Officer & Director   
498   President, Chief Executive Officer & Director   
499   President, Chief Executive Officer & Director   
500   President, Chief Executive Officer & Director   

                                     Address Ticker  \
Rank                                                  
496   8000 Bent Branch Dr., Irving, TX 75063    MIK   
497     250 Gibraltar Rd., Horsham, PA 19044    TOL   
498      701 First Ave., Sunnyvale, CA 94089   YHOO   
499         1601 Bryan St., Dallas, TX 75201    VST   
500      1 Liberty Plaza, New York, NY 10006    ABM   

                          Fullname Revenues Revchange  Profits  Prftchange  \
Rank                                                                         
496   The Michaels Companies, Inc.     5197       5.8    378.2         4.2   
497            Toll Brothers, Inc.     5170      23.9    382.1         5.2   
498                    Yahoo! Inc.     5169       4.0   -214.3         5.2   
499            Vistra Energy Corp.     5164       4.0   -214.3         5.2   
500    ABM Industries Incorporated     5145      -2.8     57.2         -25   

      Assets Totshequity  
Rank                      
496     2148     -1698.0  
497     9737      4229.0  
498    48083     31049.0  
499    15167      6597.0  
500     2281       974.0  

[5 rows x 22 columns]

In [ ]:
f500.tail(3)  #zadnje 3 vrstice

In [14]:
f500.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 1 to 500
Data columns (total 22 columns):
Title          500 non-null object
Website        500 non-null object
Employees      500 non-null int64
Sector         500 non-null object
Industry       500 non-null object
Hqlocation     500 non-null object
Hqaddr         500 non-null object
Hqcity         500 non-null object
Hqstate        500 non-null object
Hqzip          500 non-null int64
Hqtel          500 non-null object
Ceo            500 non-null object
Ceo-title      500 non-null object
Address        500 non-null object
Ticker         500 non-null object
Fullname       500 non-null object
Revenues       500 non-null int64
Revchange      500 non-null float64
Profits        500 non-null float64
Prftchange     500 non-null object
Assets         500 non-null int64
Totshequity    500 non-null float64
dtypes: float64(3), int64(4), object(15)
memory usage: 89.8+ KB


## Pandas Data Selection - indexing


In [16]:
data = pd.Series(['a','b','c'],index=[1,2,3])
data

1    a
2    b
3    c
dtype: object

In [17]:
data[1]

'a'

In [18]:
data[1:3]

2    b
3    c
dtype: object

### Selecting pandas data using “loc” (Selecting Columns From a DataFrame by Label)



<p><img alt="loc single column" src="https://s3.amazonaws.com/dq-content/291/loc_single.svg"></p>



In [20]:
# en stolpec po imenu
industries = f500.loc[:,"Industry"]
industries.head()

Rank
1                       General Merchandisers
2    Insurance: Property and Casualty (Stock)
3                 Computers, Office Equipment
4                          Petroleum Refining
5                    Wholesalers: Health Care
Name: Industry, dtype: object

In [21]:
type(industries)

pandas.core.series.Series


<p><img alt="loc list of columns" src="https://s3.amazonaws.com/dq-content/291/loc_list.svg"></p>

In [26]:
more_columns = f500.loc[:, ['Employees','Revenues','Assets']]
more_columns.head()

KeyError: 'Profits '


<p><img alt="loc slice of columns" src="https://s3.amazonaws.com/dq-content/291/loc_slice.svg"></p>

In [25]:
financial_data = f500.loc[:, 'Revenues':'Profits '] # vključno s profits
financial_data.head()

KeyError: 'Profits '

In [27]:
# Single bracket
#industries = f500.loc[:,"Industry"]
industries = f500['Industry'] # krajše stolpec
industries.head()

Rank
1                       General Merchandisers
2    Insurance: Property and Casualty (Stock)
3                 Computers, Office Equipment
4                          Petroleum Refining
5                    Wholesalers: Health Care
Name: Industry, dtype: object

In [29]:
finance = f500[['Revenues','Assets']]
finance.head()

Revenues  Assets
Rank                  
1       485873  198825
2       223604  620854
3       215639  321686
4       205004  330314
5       192487   56563

In [28]:
# dot accessor, ni najboljši način
#industries = f500.loc[:,"Industry"]
industries = f500.Industry # ta način, če ime stolpca nobenih znakov - samo ang abeceda in ni presledkov
industries.head()

Rank
1                       General Merchandisers
2    Insurance: Property and Casualty (Stock)
3                 Computers, Office Equipment
4                          Petroleum Refining
5                    Wholesalers: Health Care
Name: Industry, dtype: object


<p></p><center>
<table>
<thead>
<tr>
<th>Select by Label</th>
<th>Explicit Syntax</th>
<th>Common Shorthand</th>
<th>Other Shorthand</th>
</tr>
</thead>
<tbody>
<tr>
<td>Single column</td>
<td><code>df.loc[:,"col1"]</code></td>
<td><code>df["col1"]</code></td>
<td><code>df.col1</code></td>
</tr>
<tr>
<td>List of columns</td>
<td><code>df.loc[:,["col1", "col7"]]</code></td>
<td><code>df[["col1", "col7"]]</code></td>
<td></td>
</tr>
<tr>
<td>Slice of columns</td>
<td><code>df.loc[:,"col1":"col4"]</code></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</center><p></p>




### Selecting Items from a Series by Label



<p><img alt="dataframe exploded" src="https://s3.amazonaws.com/dq-content/291/df_exploded.svg"></p>



<p><img alt="series vs dataframe: series" src="https://s3.amazonaws.com/dq-content/291/df_series_s.svg"></p>


<p><img alt="series vs dataframe: dataframe" src="https://s3.amazonaws.com/dq-content/291/df_series_df.svg"></p>



<center>
<table>
<thead>
<tr>
<th></th>
<th>Series</th>
<th>DataFrame</th>
</tr>
</thead>
<tbody>
<tr>
<th>Dimensions</th>
<td>One</td>
<td>Two</td>
</tr>
<tr>
<th>Has 'index' axis</th>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<th>Has 'columns' axis</th>
<td>No</td>
<td>Yes</td>
</tr>
<tr>
<th>Number of dtypes</th>
<td>One</td>
<td>Many (one per column)</td>
</tr>
</tbody>
</table>
</center>



In [30]:
s = pd.Series([4,7,-5,3], index=['d', 'b', 'a', 'c'])
s

d    4
b    7
a   -5
c    3
dtype: int64

In [31]:
s['d']

4

In [32]:
s.loc['d']

4

In [33]:
s['d':'a']

d    4
b    7
a   -5
dtype: int64

In [34]:
s.loc['d':'a']

d    4
b    7
a   -5
dtype: int64

In [35]:
s[['a','c']]

a   -5
c    3
dtype: int64

In [37]:
s.loc['a':'c']

a   -5
c    3
dtype: int64

In [38]:
'a' in s

True

In [39]:
s.keys()

Index(['d', 'b', 'a', 'c'], dtype='object')

In [40]:
list(s.items())

[('d', 4), ('b', 7), ('a', -5), ('c', 3)]

In [49]:
# iz f500 izberi 
f500.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 1 to 500
Data columns (total 22 columns):
Title          500 non-null object
Website        500 non-null object
Employees      500 non-null int64
Sector         500 non-null object
Industry       500 non-null object
Hqlocation     500 non-null object
Hqaddr         500 non-null object
Hqcity         500 non-null object
Hqstate        500 non-null object
Hqzip          500 non-null int64
Hqtel          500 non-null object
Ceo            500 non-null object
Ceo-title      500 non-null object
Address        500 non-null object
Ticker         500 non-null object
Fullname       500 non-null object
Revenues       500 non-null int64
Revchange      500 non-null float64
Profits        500 non-null float64
Prftchange     500 non-null object
Assets         500 non-null int64
Totshequity    500 non-null float64
dtypes: float64(3), int64(4), object(15)
memory usage: 109.8+ KB


In [44]:
ceos = f500['Ceo'] # ali ceos = f500.Ceo ali ceos = f500.loc[:,'Ceo']
ceos
ceos.head()

Rank
1    C. Douglas McMillon
2      Warren E. Buffett
3        Timothy D. Cook
4        Darren W. Woods
5     John H. Hammergren
Name: Ceo, dtype: object

In [47]:
walmart = ceos[2] # to je dejansko index 2, št vrstice
walmart

'Warren E. Buffett'

### Selecting Rows From a DataFrame by Label



<p><img alt="anatomy of a dataframe" src="https://s3.amazonaws.com/dq-content/291/df_anatomy_static.svg"></p>



In [48]:
f500.index # od 1 do 500 nam pokaže, želimo imena kot index, ne po številki ampak po imenu iščemo

Int64Index([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,
            ...
            491, 492, 493, 494, 495, 496, 497, 498, 499, 500],
           dtype='int64', name='Rank', length=500)

**pandas.DataFrame.set_index**

http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.set_index.html?highlight=set_index#pandas.DataFrame.set_index

`DataFrame.set_index(keys, drop=True, append=False, inplace=False, verify_integrity=False)`

Set the DataFrame index (row labels) using one or more existing columns. By default yields a new object.

<table>
<colgroup><col class="field-name">
<col class="field-body">
</colgroup><tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>keys</strong> <span class="classifier-delimiter">:</span> <span class="classifier">column label or list of column labels / arrays</span></dt>
<dd></dd>
</dl>
<p><strong>drop</strong> : boolean, default True</p>
<blockquote>
<div><p>Delete columns to be used as the new index</p>
</div></blockquote>
<p><strong>append</strong> : boolean, default False</p>
<blockquote>
<div><p>Whether to append columns to existing index</p>
</div></blockquote>
<p><strong>inplace</strong> : boolean, default False</p>
<blockquote>
<div><p>Modify the DataFrame in place (do not create a new object)</p>
</div></blockquote>
<p><strong>verify_integrity</strong> : boolean, default False</p>
<blockquote>
<div><p>Check the new index for duplicates. Otherwise defer the check until
necessary. Setting to False will improve the performance of this
method</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>dataframe</strong> <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame</span></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>

In [50]:
f500.set_index('Title', inplace=True) # privzeto odvrže index - oštevilčen stolpec
f500.head(2)

Website  Employees      Sector  \
Title                                                                         
Walmart                       http://www.walmart.com    2300000   Retailing   
Berkshire Hathaway  http://www.berkshirehathaway.com     367700  Financials   

                                                    Industry       Hqlocation  \
Title                                                                           
Walmart                                General Merchandisers  Bentonville, AR   
Berkshire Hathaway  Insurance: Property and Casualty (Stock)        Omaha, NE   

                                 Hqaddr       Hqcity Hqstate  Hqzip  \
Title                                                                 
Walmart             702 S.W. Eighth St.  Bentonville      AR  72716   
Berkshire Hathaway      3555 Farnam St.        Omaha      NE  68131   

                           Hqtel  ...  \
Title                             ...   
Walmart             479-273-4000  ...   
Berkshire Hathaway  402-346-1400  ...   

                                                        Ceo-title  \
Title                                                               
Walmart             President, Chief Executive Officer & Director   
Berkshire Hathaway             Chairman & Chief Executive Officer   

                                                       Address Ticker  \
Title                                                                   
Walmart             702 S.W. Eighth St., Bentonville, AR 72716    WMT   
Berkshire Hathaway            3555 Farnam St., Omaha, NE 68131   BRKA   

                                   Fullname Revenues  Revchange  Profits  \
Title                                                                      
Walmart               Wal-Mart Stores, Inc.   485873        0.8  13643.0   
Berkshire Hathaway  Berkshire Hathaway Inc.   223604        6.1  24074.0   

                    Prftchange  Assets  Totshequity  
Title                                                
Walmart                   -7.2  198825      77798.0  
Berkshire Hathaway           0  620854     283001.0  

[2 rows x 21 columns]

In [51]:
#izbira ene vrstice
single_row = f500.loc['Walmart']
type(single_row)

pandas.core.series.Series

In [52]:
single_row

Website                               http://www.walmart.com
Employees                                            2300000
Sector                                             Retailing
Industry                               General Merchandisers
Hqlocation                                   Bentonville, AR
Hqaddr                                   702 S.W. Eighth St.
Hqcity                                           Bentonville
Hqstate                                                   AR
Hqzip                                                  72716
Hqtel                                           479-273-4000
Ceo                                      C. Douglas McMillon
Ceo-title      President, Chief Executive Officer & Director
Address           702 S.W. Eighth St., Bentonville, AR 72716
Ticker                                                   WMT
Fullname                               Wal-Mart Stores, Inc.
Revenues                                              485873
Revchange               

In [53]:
f500.head(3)

Website  Employees      Sector  \
Title                                                                         
Walmart                       http://www.walmart.com    2300000   Retailing   
Berkshire Hathaway  http://www.berkshirehathaway.com     367700  Financials   
Apple                           http://www.apple.com     116000  Technology   

                                                    Industry       Hqlocation  \
Title                                                                           
Walmart                                General Merchandisers  Bentonville, AR   
Berkshire Hathaway  Insurance: Property and Casualty (Stock)        Omaha, NE   
Apple                            Computers, Office Equipment    Cupertino, CA   

                                 Hqaddr       Hqcity Hqstate  Hqzip  \
Title                                                                 
Walmart             702 S.W. Eighth St.  Bentonville      AR  72716   
Berkshire Hathaway      3555 Farnam St.        Omaha      NE  68131   
Apple                   1 Infinite Loop    Cupertino      CA  95014   

                           Hqtel  ...  \
Title                             ...   
Walmart             479-273-4000  ...   
Berkshire Hathaway  402-346-1400  ...   
Apple               408-996-1010  ...   

                                                        Ceo-title  \
Title                                                               
Walmart             President, Chief Executive Officer & Director   
Berkshire Hathaway             Chairman & Chief Executive Officer   
Apple                          Chief Executive Officer & Director   

                                                       Address Ticker  \
Title                                                                   
Walmart             702 S.W. Eighth St., Bentonville, AR 72716    WMT   
Berkshire Hathaway            3555 Farnam St., Omaha, NE 68131   BRKA   
Apple                     1 Infinite Loop, Cupertino, CA 95014   AAPL   

                                   Fullname Revenues  Revchange  Profits  \
Title                                                                      
Walmart               Wal-Mart Stores, Inc.   485873        0.8  13643.0   
Berkshire Hathaway  Berkshire Hathaway Inc.   223604        6.1  24074.0   
Apple                           Apple, Inc.   215639       -7.7  45687.0   

                    Prftchange  Assets  Totshequity  
Title                                                
Walmart                   -7.2  198825      77798.0  
Berkshire Hathaway           0  620854     283001.0  
Apple                    -14.4  321686     128249.0  

[3 rows x 21 columns]

In [54]:
list_row = f500.loc[['Walmart', 'Berkshire Hathaway', 'Apple']]
type(list_row)

pandas.core.frame.DataFrame

In [55]:
list_row

Website  Employees      Sector  \
Title                                                                         
Walmart                       http://www.walmart.com    2300000   Retailing   
Berkshire Hathaway  http://www.berkshirehathaway.com     367700  Financials   
Apple                           http://www.apple.com     116000  Technology   

                                                    Industry       Hqlocation  \
Title                                                                           
Walmart                                General Merchandisers  Bentonville, AR   
Berkshire Hathaway  Insurance: Property and Casualty (Stock)        Omaha, NE   
Apple                            Computers, Office Equipment    Cupertino, CA   

                                 Hqaddr       Hqcity Hqstate  Hqzip  \
Title                                                                 
Walmart             702 S.W. Eighth St.  Bentonville      AR  72716   
Berkshire Hathaway      3555 Farnam St.        Omaha      NE  68131   
Apple                   1 Infinite Loop    Cupertino      CA  95014   

                           Hqtel  ...  \
Title                             ...   
Walmart             479-273-4000  ...   
Berkshire Hathaway  402-346-1400  ...   
Apple               408-996-1010  ...   

                                                        Ceo-title  \
Title                                                               
Walmart             President, Chief Executive Officer & Director   
Berkshire Hathaway             Chairman & Chief Executive Officer   
Apple                          Chief Executive Officer & Director   

                                                       Address Ticker  \
Title                                                                   
Walmart             702 S.W. Eighth St., Bentonville, AR 72716    WMT   
Berkshire Hathaway            3555 Farnam St., Omaha, NE 68131   BRKA   
Apple                     1 Infinite Loop, Cupertino, CA 95014   AAPL   

                                   Fullname Revenues  Revchange  Profits  \
Title                                                                      
Walmart               Wal-Mart Stores, Inc.   485873        0.8  13643.0   
Berkshire Hathaway  Berkshire Hathaway Inc.   223604        6.1  24074.0   
Apple                           Apple, Inc.   215639       -7.7  45687.0   

                    Prftchange  Assets  Totshequity  
Title                                                
Walmart                   -7.2  198825      77798.0  
Berkshire Hathaway           0  620854     283001.0  
Apple                    -14.4  321686     128249.0  

[3 rows x 21 columns]

In [56]:
slice_rows = f500['Berkshire Hathaway':'Exxon Mobil']
slice_rows

Website  Employees      Sector  \
Title                                                                         
Berkshire Hathaway  http://www.berkshirehathaway.com     367700  Financials   
Apple                           http://www.apple.com     116000  Technology   
Exxon Mobil                http://www.exxonmobil.com      72700      Energy   

                                                    Industry     Hqlocation  \
Title                                                                         
Berkshire Hathaway  Insurance: Property and Casualty (Stock)      Omaha, NE   
Apple                            Computers, Office Equipment  Cupertino, CA   
Exxon Mobil                               Petroleum Refining     Irving, TX   

                                    Hqaddr     Hqcity Hqstate  Hqzip  \
Title                                                                  
Berkshire Hathaway         3555 Farnam St.      Omaha      NE  68131   
Apple                      1 Infinite Loop  Cupertino      CA  95014   
Exxon Mobil         5959 Las Colinas Blvd.     Irving      TX  75039   

                           Hqtel  ...                           Ceo-title  \
Title                             ...                                       
Berkshire Hathaway  402-346-1400  ...  Chairman & Chief Executive Officer   
Apple               408-996-1010  ...  Chief Executive Officer & Director   
Exxon Mobil         972-444-1000  ...  Chairman & Chief Executive Officer   

                                                     Address Ticker  \
Title                                                                 
Berkshire Hathaway          3555 Farnam St., Omaha, NE 68131   BRKA   
Apple                   1 Infinite Loop, Cupertino, CA 95014   AAPL   
Exxon Mobil         5959 Las Colinas Blvd., Irving, TX 75039    XOM   

                                   Fullname Revenues  Revchange  Profits  \
Title                                                                      
Berkshire Hathaway  Berkshire Hathaway Inc.   223604        6.1  24074.0   
Apple                           Apple, Inc.   215639       -7.7  45687.0   
Exxon Mobil         Exxon Mobil Corporation   205004      -16.7   7840.0   

                    Prftchange  Assets  Totshequity  
Title                                                
Berkshire Hathaway           0  620854     283001.0  
Apple                    -14.4  321686     128249.0  
Exxon Mobil              -51.5  330314     167325.0  

[3 rows x 21 columns]

In [57]:
f500.head()

Website  Employees       Sector  \
Title                                                                          
Walmart                       http://www.walmart.com    2300000    Retailing   
Berkshire Hathaway  http://www.berkshirehathaway.com     367700   Financials   
Apple                           http://www.apple.com     116000   Technology   
Exxon Mobil                http://www.exxonmobil.com      72700       Energy   
McKesson                     http://www.mckesson.com      68000  Wholesalers   

                                                    Industry  \
Title                                                          
Walmart                                General Merchandisers   
Berkshire Hathaway  Insurance: Property and Casualty (Stock)   
Apple                            Computers, Office Equipment   
Exxon Mobil                               Petroleum Refining   
McKesson                            Wholesalers: Health Care   

                           Hqlocation                  Hqaddr         Hqcity  \
Title                                                                          
Walmart               Bentonville, AR     702 S.W. Eighth St.    Bentonville   
Berkshire Hathaway          Omaha, NE         3555 Farnam St.          Omaha   
Apple                   Cupertino, CA         1 Infinite Loop      Cupertino   
Exxon Mobil                Irving, TX  5959 Las Colinas Blvd.         Irving   
McKesson            San Francisco, CA              1 Post St.  San Francisco   

                   Hqstate  Hqzip         Hqtel  ...  \
Title                                            ...   
Walmart                 AR  72716  479-273-4000  ...   
Berkshire Hathaway      NE  68131  402-346-1400  ...   
Apple                   CA  95014  408-996-1010  ...   
Exxon Mobil             TX  75039  972-444-1000  ...   
McKesson                CA  94104  415-983-8300  ...   

                                                        Ceo-title  \
Title                                                               
Walmart             President, Chief Executive Officer & Director   
Berkshire Hathaway             Chairman & Chief Executive Officer   
Apple                          Chief Executive Officer & Director   
Exxon Mobil                    Chairman & Chief Executive Officer   
McKesson            Chairman, President & Chief Executive Officer   

                                                       Address Ticker  \
Title                                                                   
Walmart             702 S.W. Eighth St., Bentonville, AR 72716    WMT   
Berkshire Hathaway            3555 Farnam St., Omaha, NE 68131   BRKA   
Apple                     1 Infinite Loop, Cupertino, CA 95014   AAPL   
Exxon Mobil           5959 Las Colinas Blvd., Irving, TX 75039    XOM   
McKesson                   1 Post St., San Francisco, CA 94104    MCK   

                                   Fullname Revenues  Revchange  Profits  \
Title                                                                      
Walmart               Wal-Mart Stores, Inc.   485873        0.8  13643.0   
Berkshire Hathaway  Berkshire Hathaway Inc.   223604        6.1  24074.0   
Apple                           Apple, Inc.   215639       -7.7  45687.0   
Exxon Mobil         Exxon Mobil Corporation   205004      -16.7   7840.0   
McKesson               McKesson Corporation   192487        6.2   2258.0   

                    Prftchange  Assets  Totshequity  
Title                                                
Walmart                   -7.2  198825      77798.0  
Berkshire Hathaway           0  620854     283001.0  
Apple                    -14.4  321686     128249.0  
Exxon Mobil              -51.5  330314     167325.0  
McKesson                    53   56563       8924.0  

[5 rows x 21 columns]

In [63]:
#vaja1
top_three = f500.loc[['Walmart', 'Berkshire Hathaway', 'Apple'],['Sector','Industry']] # ali f500['Walmart':'Apple'][['Sector','Industry']]
top_three
#ali f500.loc[:,['Sector','Industry']].head(3)

SyntaxError: invalid syntax (<ipython-input-63-52260a2db466>, line 5)

In [67]:
#vaja2
#vrstice walmart:exxon mobil
#stolpci website:industry
top_five = f500['Walmart':'Exxon Mobil'].loc[:,'Website':'Industry']
top_five

Website  Employees      Sector  \
Title                                                                         
Walmart                       http://www.walmart.com    2300000   Retailing   
Berkshire Hathaway  http://www.berkshirehathaway.com     367700  Financials   
Apple                           http://www.apple.com     116000  Technology   
Exxon Mobil                http://www.exxonmobil.com      72700      Energy   

                                                    Industry  
Title                                                         
Walmart                                General Merchandisers  
Berkshire Hathaway  Insurance: Property and Casualty (Stock)  
Apple                            Computers, Office Equipment  
Exxon Mobil                               Petroleum Refining

### Selecting pandas data using “iloc”


In [68]:
#prvi dve vrstici in
top_five.iloc[0:2,:]
top_five

Website  Employees      Sector  \
Title                                                                         
Walmart                       http://www.walmart.com    2300000   Retailing   
Berkshire Hathaway  http://www.berkshirehathaway.com     367700  Financials   
Apple                           http://www.apple.com     116000  Technology   
Exxon Mobil                http://www.exxonmobil.com      72700      Energy   

                                                    Industry  
Title                                                         
Walmart                                General Merchandisers  
Berkshire Hathaway  Insurance: Property and Casualty (Stock)  
Apple                            Computers, Office Equipment  
Exxon Mobil                               Petroleum Refining

In [69]:
top_ten = f500.iloc[:10,:]
top_ten

Website  Employees  \
Title                                                             
Walmart                       http://www.walmart.com    2300000   
Berkshire Hathaway  http://www.berkshirehathaway.com     367700   
Apple                           http://www.apple.com     116000   
Exxon Mobil                http://www.exxonmobil.com      72700   
McKesson                     http://www.mckesson.com      68000   
UnitedHealth Group  http://www.unitedhealthgroup.com     230000   
CVS Health                  http://www.cvshealth.com     204000   
General Motors                     http://www.gm.com     225000   
AT&T                              http://www.att.com     268540   
Ford Motor                       http://www.ford.com     201000   

                                    Sector  \
Title                                        
Walmart                          Retailing   
Berkshire Hathaway              Financials   
Apple                           Technology   
Exxon Mobil                         Energy   
McKesson                       Wholesalers   
UnitedHealth Group             Health Care   
CVS Health                     Health Care   
General Motors      Motor Vehicles & Parts   
AT&T                    Telecommunications   
Ford Motor          Motor Vehicles & Parts   

                                                    Industry  \
Title                                                          
Walmart                                General Merchandisers   
Berkshire Hathaway  Insurance: Property and Casualty (Stock)   
Apple                            Computers, Office Equipment   
Exxon Mobil                               Petroleum Refining   
McKesson                            Wholesalers: Health Care   
UnitedHealth Group   Health Care: Insurance and Managed Care   
CVS Health          Health Care: Pharmacy and Other Services   
General Motors                      Motor Vehicles and Parts   
AT&T                                      Telecommunications   
Ford Motor                          Motor Vehicles and Parts   

                           Hqlocation                  Hqaddr         Hqcity  \
Title                                                                          
Walmart               Bentonville, AR     702 S.W. Eighth St.    Bentonville   
Berkshire Hathaway          Omaha, NE         3555 Farnam St.          Omaha   
Apple                   Cupertino, CA         1 Infinite Loop      Cupertino   
Exxon Mobil                Irving, TX  5959 Las Colinas Blvd.         Irving   
McKesson            San Francisco, CA              1 Post St.  San Francisco   
UnitedHealth Group     Minnetonka, MN        9900 Bren Rd. E.     Minnetonka   
CVS Health             Woonsocket, RI               1 CVS Dr.     Woonsocket   
General Motors            Detroit, MI  300 Renaissance Center        Detroit   
AT&T                       Dallas, TX        208 S. Akard St.         Dallas   
Ford Motor               Dearborn, MI          1 American Rd.       Dearborn   

                   Hqstate  Hqzip         Hqtel  ...  \
Title                                            ...   
Walmart                 AR  72716  479-273-4000  ...   
Berkshire Hathaway      NE  68131  402-346-1400  ...   
Apple                   CA  95014  408-996-1010  ...   
Exxon Mobil             TX  75039  972-444-1000  ...   
McKesson                CA  94104  415-983-8300  ...   
UnitedHealth Group      MN  55343  952-936-1300  ...   
CVS Health              RI   2895  401-765-1500  ...   
General Motors          MI  48265  313-556-5000  ...   
AT&T                    TX  75202  210-821-4105  ...   
Ford Motor              MI  48126  313-322-3000  ...   

                                                        Ceo-title  \
Title                                                               
Walmart             President, Chief Executive Officer & Director   
Berkshire Hathaway             Chairman & Chief Executive Officer   
Apple   

In [72]:
#top 5 vrstic vaja1
top_ten.iloc[:5,:]

Website  Employees       Sector  \
Title                                                                          
Walmart                       http://www.walmart.com    2300000    Retailing   
Berkshire Hathaway  http://www.berkshirehathaway.com     367700   Financials   
Apple                           http://www.apple.com     116000   Technology   
Exxon Mobil                http://www.exxonmobil.com      72700       Energy   
McKesson                     http://www.mckesson.com      68000  Wholesalers   

                                                    Industry  \
Title                                                          
Walmart                                General Merchandisers   
Berkshire Hathaway  Insurance: Property and Casualty (Stock)   
Apple                            Computers, Office Equipment   
Exxon Mobil                               Petroleum Refining   
McKesson                            Wholesalers: Health Care   

                           Hqlocation                  Hqaddr         Hqcity  \
Title                                                                          
Walmart               Bentonville, AR     702 S.W. Eighth St.    Bentonville   
Berkshire Hathaway          Omaha, NE         3555 Farnam St.          Omaha   
Apple                   Cupertino, CA         1 Infinite Loop      Cupertino   
Exxon Mobil                Irving, TX  5959 Las Colinas Blvd.         Irving   
McKesson            San Francisco, CA              1 Post St.  San Francisco   

                   Hqstate  Hqzip         Hqtel  ...  \
Title                                            ...   
Walmart                 AR  72716  479-273-4000  ...   
Berkshire Hathaway      NE  68131  402-346-1400  ...   
Apple                   CA  95014  408-996-1010  ...   
Exxon Mobil             TX  75039  972-444-1000  ...   
McKesson                CA  94104  415-983-8300  ...   

                                                        Ceo-title  \
Title                                                               
Walmart             President, Chief Executive Officer & Director   
Berkshire Hathaway             Chairman & Chief Executive Officer   
Apple                          Chief Executive Officer & Director   
Exxon Mobil                    Chairman & Chief Executive Officer   
McKesson            Chairman, President & Chief Executive Officer   

                                                       Address Ticker  \
Title                                                                   
Walmart             702 S.W. Eighth St., Bentonville, AR 72716    WMT   
Berkshire Hathaway            3555 Farnam St., Omaha, NE 68131   BRKA   
Apple                     1 Infinite Loop, Cupertino, CA 95014   AAPL   
Exxon Mobil           5959 Las Colinas Blvd., Irving, TX 75039    XOM   
McKesson                   1 Post St., San Francisco, CA 94104    MCK   

                                   Fullname Revenues  Revchange  Profits  \
Title                                                                      
Walmart               Wal-Mart Stores, Inc.   485873        0.8  13643.0   
Berkshire Hathaway  Berkshire Hathaway Inc.   223604        6.1  24074.0   
Apple                           Apple, Inc.   215639       -7.7  45687.0   
Exxon Mobil         Exxon Mobil Corporation   205004      -16.7   7840.0   
McKesson               McKesson Corporation   192487        6.2   2258.0   

                    Prftchange  Assets  Totshequity  
Title                                                
Walmart                   -7.2  198825      77798.0  
Berkshire Hathaway           0  620854     283001.0  
Apple                    -14.4  321686     128249.0  
Exxon Mobil              -51.5  330314     167325.0  
McKesson                    53   56563       8924.0  

[5 rows x 21 columns]

In [ ]:
#vaja2
top_ten

In [71]:
#vaja3 od 5 vrstice in 5 stolpca naprej
top_ten.iloc[6:, 6:]

Hqcity Hqstate  Hqzip         Hqtel  \
Title                                                     
CVS Health      Woonsocket      RI   2895  401-765-1500   
General Motors     Detroit      MI  48265  313-556-5000   
AT&T                Dallas      TX  75202  210-821-4105   
Ford Motor        Dearborn      MI  48126  313-322-3000   

                                  Ceo  \
Title                                   
CVS Health             Larry J. Merlo   
General Motors          Mary T. Barra   
AT&T            Randall L. Stephenson   
Ford Motor           James P. Hackett   

                                                    Ceo-title  \
Title                                                           
CVS Health      President, Chief Executive Officer & Director   
General Motors             Chairman & Chief Executive Officer   
AT&T            Chairman, President & Chief Executive Officer   
Ford Motor      President, Chief Executive Officer & Director   

                                                  Address Ticker  \
Title                                                              
CVS Health                1 CVS Dr., Woonsocket, RI 02895    CVS   
General Motors  300 Renaissance Center, Detroit, MI 48265     GM   
AT&T                   208 S. Akard St., Dallas, TX 75202      T   
Ford Motor             1 American Rd., Dearborn, MI 48126      F   

                              Fullname  Revenues  Revchange  Profits  \
Title                                                                  
CVS Health      CVS Health Corporation    177526       15.8   5317.0   
General Motors  General Motors Company    166380        9.2   9427.0   
AT&T                         AT&T Inc.    163786       11.6  12976.0   
Ford Motor          Ford Motor Company    151800        1.5   4596.0   

               Prftchange  Assets  Totshequity  
Title                                           
CVS Health            1.5   94462      36830.0  
General Motors       -2.7  221690      43836.0  
AT&T                 -2.8  403821     123135.0  
Ford Motor          -37.7  237951      29170.0

In [73]:
#vaja4 prvi stolpec in vse vrstice
top_ten.iloc[:,0] # vse vsrtice

Title
Walmart                         http://www.walmart.com
Berkshire Hathaway    http://www.berkshirehathaway.com
Apple                             http://www.apple.com
Exxon Mobil                  http://www.exxonmobil.com
McKesson                       http://www.mckesson.com
UnitedHealth Group    http://www.unitedhealthgroup.com
CVS Health                    http://www.cvshealth.com
General Motors                       http://www.gm.com
AT&T                                http://www.att.com
Ford Motor                         http://www.ford.com
Name: Website, dtype: object

In [74]:
#vaja5 10 vsrtica in vse stolpce
top_ten.iloc[9,:]


Website                                  http://www.ford.com
Employees                                             201000
Sector                                Motor Vehicles & Parts
Industry                            Motor Vehicles and Parts
Hqlocation                                      Dearborn, MI
Hqaddr                                        1 American Rd.
Hqcity                                              Dearborn
Hqstate                                                   MI
Hqzip                                                  48126
Hqtel                                           313-322-3000
Ceo                                         James P. Hackett
Ceo-title      President, Chief Executive Officer & Director
Address                   1 American Rd., Dearborn, MI 48126
Ticker                                                     F
Fullname                                  Ford Motor Company
Revenues                                              151800
Revchange               

## Series and Dataframe Describe Methods



In [75]:
#metoda describe 
f500.info()

<class 'pandas.core.frame.DataFrame'>
Index: 500 entries, Walmart to ABM Industries
Data columns (total 21 columns):
Website        500 non-null object
Employees      500 non-null int64
Sector         500 non-null object
Industry       500 non-null object
Hqlocation     500 non-null object
Hqaddr         500 non-null object
Hqcity         500 non-null object
Hqstate        500 non-null object
Hqzip          500 non-null int64
Hqtel          500 non-null object
Ceo            500 non-null object
Ceo-title      500 non-null object
Address        500 non-null object
Ticker         500 non-null object
Fullname       500 non-null object
Revenues       500 non-null int64
Revchange      500 non-null float64
Profits        500 non-null float64
Prftchange     500 non-null object
Assets         500 non-null int64
Totshequity    500 non-null float64
dtypes: float64(3), int64(4), object(14)
memory usage: 105.9+ KB


In [76]:
revs = f500['Revenues']
revs.describe() # statistični podatki za številčne podatke, preprosta analiza

count       500.000000
mean      24111.748000
std       38337.353337
min        5145.000000
25%        7245.000000
50%       11384.000000
75%       22605.250000
max      485873.000000
Name: Revenues, dtype: float64

In [77]:
f500['Assets'].describe() # serija

count    5.000000e+02
mean     8.038934e+04
std      2.704257e+05
min      4.370000e+02
25%      8.436500e+03
50%      1.932450e+04
75%      4.812600e+04
max      3.287968e+06
Name: Assets, dtype: float64

In [78]:
f500['Hqstate'].describe() # ni številk, priredi temu tipu

count     500
unique     37
top        NY
freq       54
Name: Hqstate, dtype: object

In [82]:
#cel dataframe samo za stoplce, ki imajo številčne vrednosti
f500.describe()


Employees         Hqzip       Revenues  Revchange       Profits  \
count  5.000000e+02    500.000000     500.000000  500.00000    500.000000   
mean   5.635013e+04  46791.714000   24111.748000    3.75800   1779.479800   
std    1.234520e+05  30160.384765   38337.353337   19.96748   3937.558721   
min    8.300000e+01   1104.000000    5145.000000  -57.50000  -6177.000000   
25%    1.190000e+04  19099.000000    7245.000000   -3.82500    235.725000   
50%    2.500000e+04  46244.500000   11384.000000    1.90000    683.600000   
75%    5.682525e+04  75045.000000   22605.250000    7.32500   1770.775000   
max    2.300000e+06  98188.000000  485873.000000  197.30000  45687.000000   

             Assets  Totshequity  
count  5.000000e+02     500.0000  
mean   8.038934e+04   13640.1470  
std    2.704257e+05   30523.1538  
min    4.370000e+02  -12688.0000  
25%    8.436500e+03    1997.5000  
50%    1.932450e+04    4981.0000  
75%    4.812600e+04   12467.7500  
max    3.287968e+06  283001.0000

In [ ]:
f500.describe(include = ['0'])

In [84]:
#profit
f500['Profits'].describe()

count      500.000000
mean      1779.479800
std       3937.558721
min      -6177.000000
25%        235.725000
50%        683.600000
75%       1770.775000
max      45687.000000
Name: Profits, dtype: float64

In [86]:
#revenues in employees
f500[['Revenues','Employees']].describe()

Revenues     Employees
count     500.000000  5.000000e+02
mean    24111.748000  5.635013e+04
std     38337.353337  1.234520e+05
min      5145.000000  8.300000e+01
25%      7245.000000  1.190000e+04
50%     11384.000000  2.500000e+04
75%     22605.250000  5.682525e+04
max    485873.000000  2.300000e+06

In [87]:
f500.describe(include='all')

Website     Employees      Sector  \
count                     500  5.000000e+02         500   
unique                    499           NaN          21   
top     http://www.abbott.com           NaN  Financials   
freq                        2           NaN          84   
mean                      NaN  5.635013e+04         NaN   
std                       NaN  1.234520e+05         NaN   
min                       NaN  8.300000e+01         NaN   
25%                       NaN  1.190000e+04         NaN   
50%                       NaN  2.500000e+04         NaN   
75%                       NaN  5.682525e+04         NaN   
max                       NaN  2.300000e+06         NaN   

                           Industry    Hqlocation         Hqaddr    Hqcity  \
count                           500           500            500       500   
unique                           73           246            491       239   
top     Utilities: Gas and Electric  New York, NY  390 Park Ave.  New York   
freq                             22            45              2        45   
mean                            NaN           NaN            NaN       NaN   
std                             NaN           NaN            NaN       NaN   
min                             NaN           NaN            NaN       NaN   
25%                             NaN           NaN            NaN       NaN   
50%                             NaN           NaN            NaN       NaN   
75%                             NaN           NaN            NaN       NaN   
max                             NaN           NaN            NaN       NaN   

       Hqstate         Hqzip         Hqtel  ...  \
count      500    500.000000           500  ...   
unique      37           NaN           499  ...   
top         NY           NaN  650-857-1501  ...   
freq        54           NaN             2  ...   
mean       NaN  46791.714000           NaN  ...   
std        NaN  30160.384765           NaN  ...   
min        NaN   1104.000000           NaN  ...   
25%        NaN  19099.000000           NaN  ...   
50%        NaN  46244.500000           NaN  ...   
75%        NaN  75045.000000           NaN  ...   
max        NaN  98188.000000           NaN  ...   

                                            Ceo-title  \
count                                             500   
unique                                             12   
top     President, Chief Executive Officer & Director   
freq                                              189   
mean                                              NaN   
std                                               NaN   
min                                               NaN   
25%                                               NaN   
50%                                               NaN   
75%                                               NaN   
max                                               NaN   

                                           Address Ticker  \
count                                          500    500   
unique                                         494    468   
top     601 Riverside Ave., Jacksonville, FL 32204   NFLX   
freq                                             2      3   
mean                                           NaN    NaN   
std                                            NaN    NaN   
min                                            NaN    NaN   
25%                                            NaN    NaN   
50%                                            NaN    NaN   
75%                                            NaN    NaN   
max                                            NaN    NaN   

                    Fullname       Revenues  Revchange       Profits  \
count                    500     500.000000  500.00000    500.000000   
unique                   500            NaN        NaN           NaN   
top     Amphenol Corporation            NaN        NaN           NaN   
freq                       1            NaN        NaN   

### More Data Exploration Methods



In [88]:
my_series = pd.Series([1,2,3,4,5])
my_series

0    1
1    2
2    3
3    4
4    5
dtype: int64

In [89]:
my_series = my_series + 10
my_series

0    11
1    12
2    13
3    14
4    15
dtype: int64

In [90]:
my_series.max()

15

In [91]:
my_series.min()

11

In [92]:
my_series.sum()

65

In [93]:
my_series.std()

1.5811388300841898

<p><img alt="dataframe axis parameters" src="https://s3.amazonaws.com/dq-content/291/axis_param.svg"></p>



In [95]:
df = pd.DataFrame({'A': [1,1,2,1,2,3,3,3,3,3]})
df

A
0  1
1  1
2  2
3  1
4  2
5  3
6  3
7  3
8  3
9  3

In [97]:
df.sum(axis=0) # po stolpcu 1, axis 0 vrstica

A    22
dtype: int64

In [98]:
f500.info()

<class 'pandas.core.frame.DataFrame'>
Index: 500 entries, Walmart to ABM Industries
Data columns (total 21 columns):
Website        500 non-null object
Employees      500 non-null int64
Sector         500 non-null object
Industry       500 non-null object
Hqlocation     500 non-null object
Hqaddr         500 non-null object
Hqcity         500 non-null object
Hqstate        500 non-null object
Hqzip          500 non-null int64
Hqtel          500 non-null object
Ceo            500 non-null object
Ceo-title      500 non-null object
Address        500 non-null object
Ticker         500 non-null object
Fullname       500 non-null object
Revenues       500 non-null int64
Revchange      500 non-null float64
Profits        500 non-null float64
Prftchange     500 non-null object
Assets         500 non-null int64
Totshequity    500 non-null float64
dtypes: float64(3), int64(4), object(14)
memory usage: 105.9+ KB


In [99]:
#mediana po ? # privzeto po osi 0
f500[['Revenues', 'Profits']].median(axis=0)

Revenues    11384.0
Profits       683.6
dtype: float64

In [103]:
#value counts
f500['Sector'].value_counts() # če želimo prvih 5 dodamo 

Financials                       84
Energy                           57
Retailing                        47
Technology                       43
Health Care                      38
Wholesalers                      29
Food, Beverages & Tobacco        24
Business Services                20
Industrials                      19
Materials                        19
Transportation                   17
Chemicals                        14
Engineering & Construction       13
Household Products               12
Aerospace & Defense              12
Media                            11
Telecommunications               10
Hotels, Restaurants & Leisure    10
Motor Vehicles & Parts            9
Food & Drug Stores                7
Apparel                           5
Name: Sector, dtype: int64

In [102]:
f500['Sector'].value_counts().head()

Financials     84
Energy         57
Retailing      47
Technology     43
Health Care    38
Name: Sector, dtype: int64

In [104]:
f500['Hqstate'].value_counts().head(10) #prvih 10 

NY    54
CA    53
TX    50
IL    36
OH    25
VA    23
NJ    21
PA    21
MN    18
CT    18
Name: Hqstate, dtype: int64

In [107]:
#za številčne vrednosti f500.max? se pokaže dokumentacija
f500.max(numeric_only=True)

Employees      2300000.0
Hqzip            98188.0
Revenues        485873.0
Revchange          197.3
Profits          45687.0
Assets         3287968.0
Totshequity     283001.0
dtype: float64

###  Assignment with pandas

In [109]:
my_array = np.array([1,2,3,4,5])
my_array

array([1, 2, 3, 4, 5])

In [110]:
my_array[1]=99
my_array

array([ 1, 99,  3,  4,  5])

In [113]:
top5_rank = f500[['Profits', 'Revenues']].head()
top5_rank

Profits  Revenues
Title                                
Walmart             13643.0    485873
Berkshire Hathaway  24074.0    223604
Apple               45687.0    215639
Exxon Mobil          7840.0    205004
McKesson             2258.0    192487

In [114]:
top5_rank['Revenues'] = 0
top5_rank

Profits  Revenues
Title                                
Walmart             13643.0         0
Berkshire Hathaway  24074.0         0
Apple               45687.0         0
Exxon Mobil          7840.0         0
McKesson             2258.0         0

In [115]:
#spremeni vrednost na applu, revenues
top5_rank.loc['Apple', 'Revenues'] = 999
top5_rank

Profits  Revenues
Title                                
Walmart             13643.0         0
Berkshire Hathaway  24074.0         0
Apple               45687.0       999
Exxon Mobil          7840.0         0
McKesson             2258.0         0

In [117]:
top5_rank['Year']=0 # doda stolpec in priredi 0
top5_rank

Profits  Revenues  Year
Title                                      
Walmart             13643.0         0     0
Berkshire Hathaway  24074.0         0     0
Apple               45687.0       999     0
Exxon Mobil          7840.0         0     0
McKesson             2258.0         0     0

In [118]:
top5_rank.loc['New Company'] = 555 # doda vrstico
top5_rank

Profits  Revenues  Year
Title                                      
Walmart             13643.0         0     0
Berkshire Hathaway  24074.0         0     0
Apple               45687.0       999     0
Exxon Mobil          7840.0         0     0
McKesson             2258.0         0     0
New Company           555.0       555   555

In [120]:
f500['rev_bil'] = f500['Revenues']/1000
f500.head()

Website  Employees       Sector  \
Title                                                                          
Walmart                       http://www.walmart.com    2300000    Retailing   
Berkshire Hathaway  http://www.berkshirehathaway.com     367700   Financials   
Apple                           http://www.apple.com     116000   Technology   
Exxon Mobil                http://www.exxonmobil.com      72700       Energy   
McKesson                     http://www.mckesson.com      68000  Wholesalers   

                                                    Industry  \
Title                                                          
Walmart                                General Merchandisers   
Berkshire Hathaway  Insurance: Property and Casualty (Stock)   
Apple                            Computers, Office Equipment   
Exxon Mobil                               Petroleum Refining   
McKesson                            Wholesalers: Health Care   

                           Hqlocation                  Hqaddr         Hqcity  \
Title                                                                          
Walmart               Bentonville, AR     702 S.W. Eighth St.    Bentonville   
Berkshire Hathaway          Omaha, NE         3555 Farnam St.          Omaha   
Apple                   Cupertino, CA         1 Infinite Loop      Cupertino   
Exxon Mobil                Irving, TX  5959 Las Colinas Blvd.         Irving   
McKesson            San Francisco, CA              1 Post St.  San Francisco   

                   Hqstate  Hqzip         Hqtel  ...  \
Title                                            ...   
Walmart                 AR  72716  479-273-4000  ...   
Berkshire Hathaway      NE  68131  402-346-1400  ...   
Apple                   CA  95014  408-996-1010  ...   
Exxon Mobil             TX  75039  972-444-1000  ...   
McKesson                CA  94104  415-983-8300  ...   

                                                       Address Ticker  \
Title                                                                   
Walmart             702 S.W. Eighth St., Bentonville, AR 72716    WMT   
Berkshire Hathaway            3555 Farnam St., Omaha, NE 68131   BRKA   
Apple                     1 Infinite Loop, Cupertino, CA 95014   AAPL   
Exxon Mobil           5959 Las Colinas Blvd., Irving, TX 75039    XOM   
McKesson                   1 Post St., San Francisco, CA 94104    MCK   

                                   Fullname Revenues Revchange  Profits  \
Title                                                                     
Walmart               Wal-Mart Stores, Inc.   485873       0.8  13643.0   
Berkshire Hathaway  Berkshire Hathaway Inc.   223604       6.1  24074.0   
Apple                           Apple, Inc.   215639      -7.7  45687.0   
Exxon Mobil         Exxon Mobil Corporation   205004     -16.7   7840.0   
McKesson               McKesson Corporation   192487       6.2   2258.0   

                    Prftchange  Assets Totshequity  rev_bil  
Title                                                        
Walmart                   -7.2  198825     77798.0  485.873  
Berkshire Hathaway           0  620854    283001.0  223.604  
Apple                    -14.4  321686    128249.0  215.639  
Exxon Mobil              -51.5  330314    167325.0  205.004  
McKesson                    53   56563      8924.0  192.487  

[5 rows x 22 columns]

### Using Boolean Indexing with pandas Objects



<p><img alt="Boolean arrays in pandas" src="https://s3.amazonaws.com/dq-content/291/boolean_array_pandas.svg"></p>



<p><img alt="example dataframe" src="https://s3.amazonaws.com/dq-content/291/eg_df.svg"></p>



<p><img alt="boolean series" src="https://s3.amazonaws.com/dq-content/291/bool_series.svg"></p>



<p><img alt="boolean indexing dataframe" src="https://s3.amazonaws.com/dq-content/291/boolean_indexing_df.svg"></p>



<p><img alt="boolean indexing series" src="https://s3.amazonaws.com/dq-content/291/boolean_indexing_s.svg"></p>




In [122]:
data = {'name': ['Jan', 'Jaka', 'Mihel', 'Sara'], 'num': [12,8,5,8]}
df = pd.DataFrame(data, columns=['name','num'], index=['w','x','y','z'])
df

name  num
w    Jan   12
x   Jaka    8
y  Mihel    5
z   Sara    8

In [125]:
data = {'name': ['Jan', 'Jaka', 'Mihel', 'Sara'], 'num': [12,8,5,8]}
df = pd.DataFrame(data, columns=['name','num'], index=['w','x','y','z'])
df
num_bool = df['num'] ==8
num_bool

w    False
x     True
y    False
z     True
Name: num, dtype: bool

In [126]:
df[num_bool]

name  num
x  Jaka    8
z  Sara    8

In [127]:
result = df.loc[num_bool, 'name']
result

x    Jaka
z    Sara
Name: name, dtype: object

In [129]:
#izberi iz f500 najpogostejše države z Motor Vehicles and Parts
motor_bool = f500['Industry'] == 'Motor Vehicles and Parts'
motor_bool.head(10)

Title
Walmart               False
Berkshire Hathaway    False
Apple                 False
Exxon Mobil           False
McKesson              False
UnitedHealth Group    False
CVS Health            False
General Motors         True
AT&T                  False
Ford Motor             True
Name: Industry, dtype: bool

In [131]:
motor_country = f500.loc[motor_bool, 'Hqstate']
motor_country.head()

Title
General Motors            MI
Ford Motor                MI
Lear                      MI
Goodyear Tire & Rubber    OH
Autoliv                   MI
Name: Hqstate, dtype: object

In [132]:
top5 = motor_country.value_counts().head()
top5

MI    5
OH    2
IL    1
CA    1
Name: Hqstate, dtype: int64

### Using Boolean Arrays to Assign Values



In [134]:
f500['Industry'].head(10)

Title
Walmart                                  General Merchandisers
Berkshire Hathaway    Insurance: Property and Casualty (Stock)
Apple                              Computers, Office Equipment
Exxon Mobil                                 Petroleum Refining
McKesson                              Wholesalers: Health Care
UnitedHealth Group     Health Care: Insurance and Managed Care
CVS Health            Health Care: Pharmacy and Other Services
General Motors                        Motor Vehicles and Parts
AT&T                                        Telecommunications
Ford Motor                            Motor Vehicles and Parts
Name: Industry, dtype: object

In [135]:
motor_bool = f500['Industry'] == 'Motor Vehicles and Parts'
f500.loc[motor_bool, 'Industry']= 'Motor Vehicles & Parts'
f500['Industry'].head(10)

Title
Walmart                                  General Merchandisers
Berkshire Hathaway    Insurance: Property and Casualty (Stock)
Apple                              Computers, Office Equipment
Exxon Mobil                                 Petroleum Refining
McKesson                              Wholesalers: Health Care
UnitedHealth Group     Health Care: Insurance and Managed Care
CVS Health            Health Care: Pharmacy and Other Services
General Motors                          Motor Vehicles & Parts
AT&T                                        Telecommunications
Ford Motor                              Motor Vehicles & Parts
Name: Industry, dtype: object

In [137]:
# isto skupaj
f500.loc[f500['Industry'] == 'Motor Vehicles and Parts', 'Industry']= 'Motor Vehicles & Parts'
f500['Industry'].head(10)

Title
Walmart                                  General Merchandisers
Berkshire Hathaway    Insurance: Property and Casualty (Stock)
Apple                              Computers, Office Equipment
Exxon Mobil                                 Petroleum Refining
McKesson                              Wholesalers: Health Care
UnitedHealth Group     Health Care: Insurance and Managed Care
CVS Health            Health Care: Pharmacy and Other Services
General Motors                          Motor Vehicles & Parts
AT&T                                        Telecommunications
Ford Motor                              Motor Vehicles & Parts
Name: Industry, dtype: object

### Deleting a Column from Your DataFrame


`DataFrame.drop(labels=None, axis=0, index=None, columns=None, level=None, inplace=False, errors='raise')`

Drop specified labels from rows or columns.

https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html

https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html

Remove rows or columns by specifying label names and corresponding axis, or by specifying directly index or column names. When using a multi-index, labels on different levels can be removed by specifying the level.

In [140]:
df = pd.DataFrame(data=np.array([[1, 2, 3], [40, 50, 9], [7, 2, 9], [40, 50, 9], [23, 35, 37]]), 
                  index= [2.5, 12.6, 4.8, 4.8, 2.5], 
                  columns=['A', 'B', 'C'])
df


A   B   C
2.5    1   2   3
12.6  40  50   9
4.8    7   2   9
4.8   40  50   9
2.5   23  35  37

In [143]:
df.drop(columns = 'A') #ali
df.drop('A',axis='columns', inplace=True)
df

B   C
2.5    2   3
12.6  50   9
4.8    2   9
4.8   50   9
2.5   35  37

In [144]:
df_1 = df.drop_duplicates() # privzeto prvega obdrži
df_1

B   C
2.5    2   3
12.6  50   9
4.8    2   9
2.5   35  37

### Removing a Row from Your DataFrame




`DataFrame.drop_duplicates(subset=None, keep='first', inplace=False)`

Return DataFrame with duplicate rows removed, optionally only considering certain columns

<table class="docutils field-list" frame="void" rules="none">
<colgroup><col class="field-name">
<col class="field-body">
</colgroup><tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>subset</strong> : column label or sequence of labels, optional</p>
<blockquote>
<div><p>Only consider certain columns for identifying duplicates, by
default use all of the columns</p>
</div></blockquote>
<p><strong>keep</strong> : {‘first’, ‘last’, False}, default ‘first’</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">first</span></code> : Drop duplicates except for the first occurrence.</li>
<li><code class="docutils literal notranslate"><span class="pre">last</span></code> : Drop duplicates except for the last occurrence.</li>
<li>False : Drop all duplicates.</li>
</ul>
</div></blockquote>
<p><strong>inplace</strong> : boolean, default False</p>
<blockquote>
<div><p>Whether to drop duplicates in place or to return a copy</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>deduplicated</strong> <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame</span></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>

In [145]:
df_1.drop([2.5])

B  C
12.6  50  9
4.8    2  9

### Combining Datasets: Concat and Append


In [4]:
def make_df(cols, ind):
    """Quickly make a DataFrame"""
    data = {c: [str(c) + str(i) for i in ind]
            for c in cols}
    return pd.DataFrame(data, ind)



In [5]:
# example DataFrame
make_df('ABCD', range(3))

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2

#### Concatenation of NumPy Arrays



In [6]:
x = [1, 2, 3]
y = [4, 5, 6]
z = [7, 8, 9]


#### Simple Concatenation with pd.concat



`pandas.concat(objs, axis=0, join='outer', join_axes=None, ignore_index=False, keys=None, levels=None, names=None, verify_integrity=False, sort=None, copy=True)`

[Dokumentacija](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html)


In [7]:
ser1 = pd.Series(['A', 'B', 'C'], index=[1, 2, 3])
ser2 = pd.Series(['D', 'E', 'F'], index=[4, 5, 6])


In [8]:
df1 = make_df('AB', [1, 2])
df2 = make_df('AB', [3, 4])



In [56]:
df3 = make_df('AB', [0, 1])
df4 = make_df('CD', [0, 1])

#### Duplicate indices



In [61]:
x = make_df('AB', [0, 1])
y = make_df('AB', [2, 3])

##### Catching the repeats as an error

##### Ignoring the index

##### Adding MultiIndex keys

[Več o Hierarchical Indexing](https://jakevdp.github.io/PythonDataScienceHandbook/03.05-hierarchical-indexing.html) 

#### Concatenation with joins

In [74]:
df5 = make_df('ABC', [1, 2])
df6 = make_df('BCD', [3, 4])

#### The append() method



### Aggregation and Grouping

#### Planets dataset



#### GroupBy: Split, Apply, Combine

##### Split, apply, combine

<img src="https://jakevdp.github.io/PythonDataScienceHandbook/figures/03.08-split-apply-combine.png" alt="">



In [9]:
df = pd.DataFrame({'key': ['A', 'B', 'C', 'A', 'B', 'C'],
                   'data': range(6)}, columns=['key', 'data'])


#### The GroupBy object

#### Column indexing

#### Dispatch methods



#### Aggregate, filter, transform, apply

In [10]:
rng = np.random.RandomState(0)
df = pd.DataFrame({'key': ['A', 'B', 'C', 'A', 'B', 'C'],
                   'data1': range(6),
                   'data2': rng.randint(0, 10, 6)},
                   columns = ['key', 'data1', 'data2'])


##### Aggregation


##### Filtering



##### Transformation

##### The apply() method

### Working with Time Series

#### Dates and Times in Python


##### Native Python dates and times



##### Typed arrays of times: NumPy's datetime64

[Dokumentacija](https://docs.scipy.org/doc/numpy/reference/arrays.datetime.html)

<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<table>
<thead><tr>
<th>Code</th>
<th>Meaning</th>
<th>Time span (relative)</th>
<th>Time span (absolute)</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>Y</code></td>
<td>Year</td>
<td>± 9.2e18 years</td>
<td>[9.2e18 BC, 9.2e18 AD]</td>
</tr>
<tr>
<td><code>M</code></td>
<td>Month</td>
<td>± 7.6e17 years</td>
<td>[7.6e17 BC, 7.6e17 AD]</td>
</tr>
<tr>
<td><code>W</code></td>
<td>Week</td>
<td>± 1.7e17 years</td>
<td>[1.7e17 BC, 1.7e17 AD]</td>
</tr>
<tr>
<td><code>D</code></td>
<td>Day</td>
<td>± 2.5e16 years</td>
<td>[2.5e16 BC, 2.5e16 AD]</td>
</tr>
<tr>
<td><code>h</code></td>
<td>Hour</td>
<td>± 1.0e15 years</td>
<td>[1.0e15 BC, 1.0e15 AD]</td>
</tr>
<tr>
<td><code>m</code></td>
<td>Minute</td>
<td>± 1.7e13 years</td>
<td>[1.7e13 BC, 1.7e13 AD]</td>
</tr>
<tr>
<td><code>s</code></td>
<td>Second</td>
<td>± 2.9e12 years</td>
<td>[ 2.9e9 BC, 2.9e9 AD]</td>
</tr>
<tr>
<td><code>ms</code></td>
<td>Millisecond</td>
<td>± 2.9e9 years</td>
<td>[ 2.9e6 BC, 2.9e6 AD]</td>
</tr>
<tr>
<td><code>us</code></td>
<td>Microsecond</td>
<td>± 2.9e6 years</td>
<td>[290301 BC, 294241 AD]</td>
</tr>
<tr>
<td><code>ns</code></td>
<td>Nanosecond</td>
<td>± 292 years</td>
<td>[ 1678 AD, 2262 AD]</td>
</tr>
<tr>
<td><code>ps</code></td>
<td>Picosecond</td>
<td>± 106 days</td>
<td>[ 1969 AD, 1970 AD]</td>
</tr>
<tr>
<td><code>fs</code></td>
<td>Femtosecond</td>
<td>± 2.6 hours</td>
<td>[ 1969 AD, 1970 AD]</td>
</tr>
<tr>
<td><code>as</code></td>
<td>Attosecond</td>
<td>± 9.2 seconds</td>
<td>[ 1969 AD, 1970 AD]</td>
</tr>
</tbody>
</table>

</div>
</div>

#### Dates and times in pandas: best of both worlds


#### Pandas Time Series: Indexing by Time

#### Pandas Time Series Data Structures


In [13]:
dates = pd.to_datetime([datetime(2015, 7, 3), '4th of July, 2015',
                       '2015-Jul-6', '07-07-2015', '20150708'])



#### Example

### Understanding SettingwithCopyWarning in pandas


#### What is SettingWithCopyWarning?



<img class="full-width" src="https://www.dataquest.io/wp-content/uploads/2019/01/view-vs-copy.png" alt="view-vs-copy">



<img class="full-width" src="https://www.dataquest.io/wp-content/uploads/2019/01/modifying.png" alt="modifying">



## Več o 

[Pandas on PyVideo](https://pyvideo.org/)